types/ssri/index.d.ts

```diff
@@ -1,126 +1,181 @@
 
 import { Utf8AsciiLatin1Encoding, Hash as CryptoHash } from "crypto";
 import { Readable, Transform } from "stream";
 
 export interface HashLike {
-    algorithm: string;
-    digest: string;
-    options?: string[];
+	algorithm: string;
+	digest: string;
+	options?: string[];
 }
 
 export interface IntegrityLike {
-    [algorithm: string]: HashLike[];
+	[algorithm: string]: HashLike[];
 }
 
 export type IntegrityMap = Integrity & IntegrityLike;
 
 export class Hash implements HashLike {
-    constructor(hash: string, opts?: { strict?: boolean });
+	constructor(hash: string, opts?: { strict?: boolean });
 
-    source: string;
-    algorithm: string;
-    digest: string;
-    options?: string[];
-    isHash: boolean;
+	source: string;
+	algorithm: string;
+	digest: string;
+	options?: string[];
+	isHash: boolean;
 
-    hexDigest(): string;
+	hexDigest(): string;
 
-    toJSON(): string;
+	toJSON(): string;
 
-    toString(opts?: { strict?: boolean }): string;
+	toString(opts?: { strict?: boolean }): string;
 }
 
 export class Integrity {
-    isIntegrity: boolean;
+	isIntegrity: boolean;
 
-    toJSON(): string;
+	toJSON(): string;
 
-    toString(opts?: { strict?: boolean, sep?: string }): string;
+	toString(opts?: { strict?: boolean, sep?: string }): string;
 
-    concat(
-        integrity: string | IntegrityLike | HashLike,
-        opts?: { strict?: boolean }
-    ): IntegrityMap;
+	concat(
+		integrity: string | IntegrityLike | HashLike,
+		opts?: { strict?: boolean },
+	): IntegrityMap;
 
-    hexDigest(): string;
+	hexDigest(): string;
 
-    match(
-        integrity: string | IntegrityLike | HashLike,
-        opts?: { strict?: boolean, pickAlgorithm?: (algo1: string, algo2: string) => string }
-    ): Hash | false;
+	match(
+		integrity: string | IntegrityLike | HashLike,
+		opts?: {
+			strict?: boolean,
+			pickAlgorithm?: (algo1: string, algo2: string) => string,
+		},
+	): Hash | false;
 
-    pickAlgorithm(opts?: { pickAlgorithm?: (algo1: string, algo2: string) => string }): string;
+	pickAlgorithm(opts?: {
+		pickAlgorithm?: (algo1: string, algo2: string) => string,
+	}): string;
 }
 
 export function parse(
-    sri: string | IntegrityLike | HashLike,
-    opts?: { strict?: boolean }
+	sri: string | IntegrityLike | HashLike,
+	opts?: { single?: false, strict?: boolean, },
 ): IntegrityMap;
-
 export function parse(
-    sri: string | IntegrityLike | HashLike,
-    opts?: { single: true, strict?: boolean }
+	sri: string | IntegrityLike | HashLike,
+	opts?: { single: true, strict?: boolean, },
 ): Hash;
+export function parse(
+	sri: string | IntegrityLike | HashLike,
+	opts?: { single?: boolean, strict?: boolean, },
+): IntegrityMap | Hash;
 
 export function stringify(
-    obj: string | IntegrityLike | HashLike,
-    opts?: { strict?: boolean, sep?: string }
+	obj: string | IntegrityLike | HashLike,
+	opts?: { strict?: boolean, sep?: string, },
 ): string;
 
 export function fromHex(
```

Comment:
Another option is to just have one function like:
```
export function fromHex<T extends boolean = false>(
  hexDigest: string, 
  algorithm: string, 
  opts?: { 
    single?: T, 
    string?: boolean, 
    options?: ReadonlyArray<string>
  }
): T extends true ? Hash : IntegrityMap;
```

Reply:
That would bump the required TypeScript version to 2.8.

Reply:
The TypeScript version should really be specified in the top of the d.ts file if people are going to be relying on a fixed version.